homeassistant/components/stream/hls.py

```diff
@@ -113,47 +156,137 @@ def render(track: StreamOutput) -> str:
             # don't autoplay. Also, hls.js uses the player parameter liveSyncDuration
             # which seems to take precedence for setting target delay. Yet it also
             # doesn't seem to hurt, so we can stick with it for now.
-            f"#EXT-X-START:TIME-OFFSET=-{EXT_X_START * track.target_duration:.3f}",
-        ]
+            playlist.append(
+                f"#EXT-X-START:TIME-OFFSET=-{EXT_X_START_NON_LL_HLS*track.target_duration:.3f},PRECISE=YES"
+            )
 
         last_stream_id = first_segment.stream_id
-        # Add playlist sections
-        for segment in segments:
-            # Skip last segment if it is not complete
-            if segment.complete:
-                if last_stream_id != segment.stream_id:
-                    playlist.extend(
-                        [
-                            "#EXT-X-DISCONTINUITY",
-                            "#EXT-X-PROGRAM-DATE-TIME:"
-                            + segment.start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
-                            + "Z",
-                        ]
-                    )
-                playlist.extend(
-                    [
-                        f"#EXTINF:{segment.duration:.3f},",
-                        f"./segment/{segment.sequence}.m4s",
-                    ]
+
+        # Add playlist sections for completed segments
+        # Enumeration used to only include EXT-X-PART data for last 3 segments.
+        # The RFC seems to suggest removing parts after 3 full segments, but Apple's
+        # own example shows removing after 2 full segments and 1 part one.
+        for i, segment in enumerate(segments[:-1], 3 - len(segments)):
+            playlist.append(
+                segment.render_hls(
+                    last_stream_id=last_stream_id, render_parts=i >= 0, add_hint=False
```

Comment:
I'm having trouble seeing what is happening with `render_parts` here.  Is `i` always positive here?

Reply:
The enumeration starts from `3-len(segments)`, so with the playlist size we are currently using (3 full segments + 1 part segment), `i` will generally be negative for the first segment. The effects of this would probably be more clear if we were using longer playlists with say 10 segments. We can consider simplifying it if we are sure we are sticking with the current playlist size, although I'm not sure if any simplification will be much simpler, maybe just less confusing.